testresultparser/parser.go

```diff
@@ -0,0 +1,180 @@
+		Time:      f.TimeOut,
+	}
+	result.Failure = append(result.Failure,
+		Failure{TestCase: "",
+			Type:    "",
+			Message: f.FailureMessage,
+		})
+	return result
+}
+
+// convertResultFromXML is a helper function
+// to convert struct to desired output struct
+func convertResultFromXML(t TestResultXML) Result {
+	result := Result{
+		TestSuite: t.TestSuite,
+		Tests:     t.Tests,
+		Failures:  t.Failures,
+		Errors:    t.Errors,
+		Skipped:   t.Skipped,
+		Time:      t.Time}
+
+	for _, test := range t.T {
+		if test.F.FailureType != "" {
+			result.Failure = append(result.Failure,
+				Failure{TestCase: test.Name,
+					Type:    test.F.FailureType,
+					Message: test.F.Message,
+				})
+		}
+	}
+	return result
+}
+
+// Parse method for Text Report parsing
+func (r *Result) Parse(f []byte) *Result {
+	result := Result{}
+
+	testsuite := regexp.MustCompile("Test set: ([a-zA-Z0-9_.]+)")
+	tests := regexp.MustCompile("Tests run: ([a-zA-Z0-9_.]+)")
+	failures := regexp.MustCompile("Failures: ([a-zA-Z0-9_.]+)")
+	errors := regexp.MustCompile("Errors: ([a-zA-Z0-9_.]+)")
+	skipped := regexp.MustCompile("Skipped: ([a-zA-Z0-9_.]+)")
+	time := regexp.MustCompile("Time elapsed: ([a-zA-Z0-9_.]+)")
+
+	result.TestSuite = testsuite.FindStringSubmatch(string(f))[1]
+	result.Tests, _ = strconv.Atoi(tests.FindStringSubmatch(string(f))[1])
+	result.Failures, _ = strconv.Atoi(failures.FindStringSubmatch(string(f))[1])
+	result.Errors, _ = strconv.Atoi(errors.FindStringSubmatch(string(f))[1])
+	result.Skipped, _ = strconv.Atoi(skipped.FindStringSubmatch(string(f))[1])
+	result.Time = time.FindStringSubmatch(string(f))[1]
+
+	if result.Failures != 0 {
+		failedTestCase := regexp.MustCompile(`(.*)\(.*\).*FAILURE`)
+		failureType := regexp.MustCompile(`(.*Error): (.*)`)
+		for i := 0; i < result.Failures; i++ {
+			result.Failure = append(result.Failure,
+				Failure{TestCase: failedTestCase.FindAllStringSubmatch(string(f), -1)[i][1],
+					Type:    failureType.FindAllStringSubmatch(string(f), -1)[i][1],
+					Message: failureType.FindAllStringSubmatch(string(f), -1)[i][2],
+				})
+		}
+	}
+	return &result
+}
+
+// Parse method for XML Report parsing
+func (tr *TestResultXML) Parse(b []byte) *Result {
+	xml.Unmarshal(b, tr)
+	result := convertResultFromXML(*tr)
+	return &result
+}
+
+// Parse method for Failsafe XML Report parsing
+func (fr *failsafeReportXML) Parse(b []byte) *Result {
+	xml.Unmarshal(b, fr)
+	result := convertResultFromFailsafe(*fr)
+	return &result
+}
+
+// Parser interface to buildtools parse methods
+type Parser interface {
+	Parse([]byte) *Result
+}
+
+// ReportParser returns byte object to the interface parse method
+func ReportParser(p Parser, filepath string) *Result {
+	fp, _ := os.Open(filepath)
+	defer fp.Close()
+	b, _ := ioutil.ReadAll(fp)
+	r := p.Parse(b)
+	return r
+}
+
+// Parse function is called by the external interface and
+// returns the output struct
+func Parse(filepath, plugin string) (*Result, error) {
```

Comment:
This will make sense if those files have some pattern indicating the "source system" (Gradle, Failsafe, Surefire), otherwise we can leave it with inverted control.